controllers/stucknodecondition/stuck_node_checker.go

```diff
@@ -0,0 +1,341 @@
+package upgradeconfig
+
+import (
+	"context"
+	"fmt"
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	policyv1 "k8s.io/api/policy/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/utils/clock"
+	"sort"
+	"strings"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+var (
+	log = logf.Log.WithName("controller_stucknode")
+)
+
+// blank assignment to verify that StuckNode implements reconcile.Reconciler
+var _ reconcile.Reconciler = &StuckNode{}
+
+// StuckNode reconciles the StuckNode condition.
+// This condition is intentionally independent of the upgrade itself since nodes can be stuck during any drain, regardless
+// of which upgrade or any upgrade in progress.
+type StuckNode struct {
+	dynamicClient        dynamic.Interface
+	kubeClient           kubernetes.Interface
+	ongoingDrains        map[string]time.Time
+	drainTimeoutDuration time.Duration
+	clock                clock.Clock
+}
+
+func (r *StuckNode) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	reqLogger := log.WithValues("Request.Namespace", request.Namespace, "Request.Name", request.Name)
+	reqLogger.Info("Reconciling stuck node checker")
+	// TODO seems much cleaner to wire up logger in context like kube
+
+	reconcileErr := r.reconcile(ctx, reqLogger)
+	if reconcileErr != nil {
+		// use real generated client to set only the StuckNodeControllerDegraded condition
+		return reconcile.Result{}, reconcileErr
+	}
+
+	// use real generated client to set only the StuckNodeControllerDegraded condition
+
+	return reconcile.Result{}, reconcileErr
+}
+
+func (r *StuckNode) reconcile(ctx context.Context, logger logr.Logger) error {
+	allNodes, err := r.kubeClient.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to list nodes: %w", err)
+	}
+
+	drainingNodes := []corev1.Node{}
+	drainingNodesByName := map[string]corev1.Node{}
+	for _, node := range allNodes.Items {
+		// draining nodes are unschedulable
+		if !node.Spec.Unschedulable {
+			continue
+		}
+		// the MCO uses this annotation with a suffix of drain to indicate draining.
+		if !strings.HasSuffix(node.Annotations["machineconfiguration.openshift.io/desiredDrain"], "drain") {
```

Comment:
Checking in case this is too new for our management clusters:

```console
$ for Y in 10 11; do git --no-pager grep 'machineconfiguration.openshift.io/desiredDrain' "origin/release-4.${Y}"; done
origin/release-4.11:pkg/daemon/constants/constants.go:  DesiredDrainerAnnotationKey = "machineconfiguration.openshift.io/desiredDrain"
```

So all 4.(y>=11) management clusters should have an MCO that sets this annotation, and our management clusters are all 4.15 or 4.16 or newer or something.  So should be fine as you have it, but commenting in case other folks also wonder about this.